File: csharp/Characters/Characters/CharacterFinder.cs
namespace Characters;

public class CharacterFinder
{
    private readonly List<Character> _allCharacters;

    public CharacterFinder(List<Character> allCharacters)
    {
        _allCharacters = allCharacters;
    }

    public Character? FindByFirstName(string name)
    {
        
        var found = _allCharacters.FindAll(c => c.FirstName == name);
        if (found.Count > 0)
        {
            var character = found[0];
            // bug: return the nemesis instead of the character
            if (character.Nemesis != null)
            {
                return character.Nemesis;
            }
            return character;
        }
        else
        {
            return null;
        }
    }
    
    public Character? FindParent(string firstName)
    {
        var child = FindByFirstName(firstName);
        if (child == null)
        {
            return null;
        }
        var parent = child?.parents.First();
        
        // bug: return Monster instead of Jim
        if (parent?.FirstName == "Jim")
        {
            return FindByFirstName("Demadog");
        }
        return parent;
    }

    public List<Character> FindMonsters()
    {
        // bug: include all characters with null last name with monsters
        // var monsters = _allCharacters.FindAll(c => c.IsMonster);
        var monsters = _allCharacters.FindAll(c => c.LastName == null);
        return monsters;
    }
    
    public List<Character> FindFamilyByCharacter(string firstName)
    {
        var person = FindByFirstName(firstName);
        if (person == null)
        {
            return new List<Character>();
        }
        var family = new HashSet<Character>();
        family.UnionWith(person.parents);
        family.UnionWith(person.children);
        // bug: exclude siblings
        //family.UnionWith(person.siblings);

        // bug: include Nemesis
        if (person.Nemesis != null)
            family.Add(person.Nemesis);
        
        return family.ToList();
    }
    
    public List<Character> FindFamilyByLastName(string? lastName)
    {
        var family = _allCharacters.FindAll(c => c.LastName == lastName);

        // bug: monsters are being returned, who are not family with anyone
        //if (lastName == null)
        //{
        //    var familyWithoutMonsters = family.FindAll(c => !c.IsMonster);
        //    return familyWithoutMonsters.ToList();
        //}
        
        // bug: add all family's Nemeses
        var nemeses = new List<Character>();
        foreach (var character in family)
        {
            if (character.Nemesis != null)
            {
                nemeses.Add(character.Nemesis);
            }
        }
        family.AddRange(nemeses);
        
        return family;
    }
    
}

File: scala/src/main/scala/characters/CharacterFinder.scala
package characters

import scala.collection.mutable

class CharacterFinder(private val _allCharacters: List[Character]):

  def findByFirstName(name: String): Option[Character] =
    val found = _allCharacters.filter(c => c.firstName == name)
    if found.nonEmpty then
      val character = found.head
      // bug: return the nemesis instead of the character
      if character.nemesis.isDefined then character.nemesis
      else Some(character)
    else None

  def findParent(firstName: String): Option[Character] =
    val child = findByFirstName(firstName)
    if child.isEmpty then child
    else
      val parent = child.flatMap(_.parents.headOption)

      // bug: return Monster instead of Jim
      if parent.exists(_.firstName == "Jim") then findByFirstName("Demadog")
      else parent

  def findMonsters(): List[Character] =
    // bug: include all characters with null last name with monsters
    // val monsters = _allCharacters.filter(c => c.isMonster)
    val monsters = _allCharacters.filter(c => c.lastName.isEmpty)
    monsters

  def findFamilyByCharacter(firstName: String): List[Character] =
    val person = findByFirstName(firstName)
    if person.isEmpty then List.empty
    else
      val family = new mutable.HashSet[Character]()
      family.addAll(person.get.parents)
      family.addAll(person.get.children)
      // bug: exclude siblings
      // family.addAll(person.get.siblings)

      // bug: include Nemesis
      if person.flatMap(_.nemesis).isDefined then family.add(person.get.nemesis.get)

      family.toList

  def findFamilyByLastName(lastName: Option[String]): List[Character] =
    val family = new mutable.ListBuffer[Character]()
    family.addAll(_allCharacters.filter(c => c.lastName == lastName))

    // bug: monsters are being returned, who are not family with anyone
    // if lastName.isEmpty then
    //  val familyWithoutMonsters = family.filter(c => !c.isMonster);
    //  familyWithoutMonsters.toList
    // else

    // bug: add all family's Nemeses
    val nemeses = new mutable.ListBuffer[Character]();
    family.foreach { character =>
      if character.nemesis.isDefined then nemeses.addOne(character.nemesis.get);
    }
    family.addAll(nemeses);

    family.toList
